src/sync/app.cpp

```diff
@@ -87,9 +89,32 @@ const static uint64_t    default_timeout_ms = 60000;
 const static std::string username_password_provider_key = "local-userpass";
 const static std::string user_api_key_provider_key_path = "api_keys";
 
-SharedApp App::get_shared_app(const Config& config)
+SharedApp App::get_shared_app(const Config& config, const SyncClientConfig& sync_client_config)
 {
-    return std::make_shared<App>(config);
+    if (m_apps_cache.find(config.app_id) != m_apps_cache.end()) {
+        return m_apps_cache[config.app_id];
+    }
+
+    auto app = std::make_shared<App>(config);
+    app->configure(sync_client_config);
+    m_apps_cache[config.app_id] = app;
+    return app;
```

Comment:
This is doing a whole bunch of lookups in the map when it only needs to be doing one:

```suggestion
    auto& app = m_apps_cache[config.app_id];
    if (!app) {
        app = std::make_shared<App>(config);
        app->configure(sync_client_config);
    }
    return app;
```